### PR TITLE
Benchmark performance for HarbourBridge migration

### PR DESCRIPTION
**Pull request by shreyakhajanchi**
Changes part of this PR:
- Script to create and delete instance
- Populate single table database with random data
- New flag for specifying the write limit
- Computing the duration of schema and data conversion and populating the same in report

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -36,6 +36,8 @@ var (
 	sessionFile = "session.json"
 )
 
+const defaultWriteLimit = 40
```
Comment (bharadwaj-aditya) [inline cmd/cmd.go:39]: Nit: this is the number of writers. Suggest we rename to defaultWritersLimit.

File: cmd/data.go [hunk 2]
```diff
@@ -57,6 +58,7 @@ func (cmd *DataCmd) SetFlags(f *flag.FlagSet) {
 	f.StringVar(&cmd.targetProfile, "target-profile", "", "Flag for specifying connection profile for target database e.g., \"dialect=postgresql\"")
 	f.BoolVar(&cmd.skipForeignKeys, "skip-foreign-keys", false, "Skip creating foreign keys after data migration is complete (ddl statements for foreign keys can still be found in the downloaded schema.ddl.txt file and the same can be applied separately)")
 	f.StringVar(&cmd.filePrefix, "prefix", "", "File prefix for generated files")
+	f.Int64Var(&cmd.writeLimit, "write-limit", 40, "Write limit for writes to spanner")
```
Comment (Deep1998) [inline cmd/data.go:61]: The default value here should be the constants variable.

File: performance/setup.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        echo "$insertRecordCount records inserted."
+        for nodes in 1 3 5
+        do
+            gcloud spanner instances create new-test-instance --config=regional-us-central1 \
+            --description="New test Instance" --nodes=$nodes
+            gcloud config set spanner/instance new-test-instance
+            echo "Spanner instance created with $nodes nodes"
+            for writeLimit in 20 40 60 80 100
+            do
+                echo "Iteration $threads"
```
Comment (Deep1998) [inline performance/setup.sh:23]: Threads is not defined before this

**Review by Deep1998 — COMMENTED** [review]
LGTM, waiting for Aditya's review

File: internal/report.go [hunk 4]
```diff
@@ -557,3 +558,18 @@ func writeHeading(w *bufio.Writer, s string) {
 		s, "\n",
 		"----------------------------\n"}, ""))
 }
+
+func conversionDuration(conv *Conv, w *bufio.Writer) string {
+	res := ""
+	if conv.DataConversionDuration.Microseconds() != 0 || conv.SchemaConversionDuration.Microseconds() != 0 {
```
Comment (bharadwaj-aditya) [inline internal/report.go:564]: why is the check for non zero required ? We should be able to print it in any case.

File: performance/cleanup_resource/cleanup_resource.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"database/sql"
+	"log"
+	"os"
+
+	_ "github.com/go-sql-driver/mysql"
+
+	"github.com/cloudspannerecosystem/harbourbridge/performance"
+)
+
```
Comment (bharadwaj-aditya) [inline performance/cleanup_resource/cleanup_resource.go:12]: Please add a doc/comment on the purpose of this module

File: performance/populate_database/populate_database.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	"flag"
+	"log"
+	"os"
+	"strconv"
+
+	"github.com/cloudspannerecosystem/harbourbridge/performance"
+	"github.com/go-sql-driver/mysql"
+)
+
+const (
```
Comment (bharadwaj-aditya) [inline performance/populate_database/populate_database.go:15]: Please add some docs here outlining
1. Purpose of the module
2. How and when to execute it.

File: performance/setup.sh [hunk 7]
```diff
@@ -0,0 +1,53 @@
+#!/bin/sh
+if type mysql >/dev/null 2>&1; then
+    echo "MySQL present."
+    export MYSQLHOST=localhost
+    export MYSQLUSER=root
+    export MYSQLPORT=3306
+    #update MySQL password before running the benchmark
+    export MYSQLPWD=
+    gcloud auth application-default login
```
Comment (bharadwaj-aditya) [inline performance/setup.sh:9]: Nit: Add a print statement after this that login was successful and records are going to be inserted

File: performance/util.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
+package performance
+
+import (
+	"fmt"
+	"math/rand"
+	"strings"
+	"time"
+)
+
+func RandomInt(min, max int64) int64 {
```
Comment (bharadwaj-aditya) [inline performance/util.go:10]: Please add docs on the purpose of this module and what should be added here

**Review by bharadwaj-aditya — CHANGES_REQUESTED** [review]
Minor comments, mostly around documentation

File: performance/setup.sh (hunk 7 above)
Comment (shreyakhajanchi) [reply performance/setup.sh:9]: added

File: performance/populate_database/populate_database.go (hunk 6 above)
Comment (shreyakhajanchi) [reply performance/populate_database/populate_database.go:15]: Sure, makes sense added a readme file

File: performance/cleanup_resource/cleanup_resource.go (hunk 5 above)
Comment (shreyakhajanchi) [reply performance/cleanup_resource/cleanup_resource.go:12]: added comments

**Review by bharadwaj-aditya — APPROVED** [review]
LGTM

**Pull request merged**